# 📑 Pipeline de Inteligência de Dados: Churn de Clientes TelecomX Parte 2

**Contexto:** Este notebook documenta a pipeline de Ciência de Dados desenvolvida para a **TelecomX**. O projeto foca no problema de `Churn` (cancelamento de clientes), utilizando uma abordagem que vai da análise exploratória profunda até a implementação de modelos preditivos avançados.

**Objetivo:** Identificar os perfis de risco, compreender as causas financeiras e contratuais da evasão e selecionar o melhor algoritmo de Machine Learning para antecipar cancelamentos e apoiar a tomada de decisão estratégica.

---

<br/>
<img src="https://raw.githubusercontent.com/WBruni/analise-churn-telecom-x_parte_2/main/images/wokflow.png" width="80%">

# 📌 Extração (Coleta)

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/WBruni/analise-churn-telecom-x_parte_2/refs/heads/main/data/dados_tratados.csv')

In [3]:
df.head()

,CustomerID,Churn,Gender,SeniorCitizen,Partner,Dependents,Tenure,PhoneService,MultipleLines,InternetService,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,ChargesDaily,ChargesMonthly,ChargesTotal
0,0002-ORFBO,No,Female,No,Yes,Yes,9,Yes,No,DSL,...,No,Yes,Yes,No,One year,Yes,Mailed check,2.19,65.6,593.30
1,0003-MKNFE,No,Male,No,No,No,9,Yes,Yes,DSL,...,No,No,No,Yes,Month-to-month,No,Mailed check,2.00,59.9,542.40
2,0004-TLHLJ,Yes,Male,No,No,No,4,Yes,No,Fiber optic,...,Yes,No,No,No,Month-to-month,Yes,Electronic check,2.46,73.9,280.85
3,0011-IGKFF,Yes,Male,Yes,Yes,No,13,Yes,No,Fiber optic,...,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,3.27,98.0,1237.85
4,0013-EXCHZ,Yes,Female,Yes,Yes,No,3,Yes,No,Fiber optic,...,No,Yes,Yes,No,Month-to-month,Yes,Mailed check,2.80,83.9,267.40


In [4]:
df_raw = df.copy()

In [5]:
df = df.drop(columns=['CustomerID'])

In [6]:
df = df = df.drop_duplicates()

In [7]:
df.columns

Index(['Churn', 'Gender', 'SeniorCitizen', 'Partner', 'Dependents', 'Tenure',
       'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity',
       'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV',
       'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod',
       'ChargesDaily', 'ChargesMonthly', 'ChargesTotal'],
      dtype='object')

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7021 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Churn             7021 non-null   object 
 1   Gender            7021 non-null   object 
 2   SeniorCitizen     7021 non-null   object 
 3   Partner           7021 non-null   object 
 4   Dependents        7021 non-null   object 
 5   Tenure            7021 non-null   int64  
 6   PhoneService      7021 non-null   object 
 7   MultipleLines     7021 non-null   object 
 8   InternetService   7021 non-null   object 
 9   OnlineSecurity    7021 non-null   object 
 10  OnlineBackup      7021 non-null   object 
 11  DeviceProtection  7021 non-null   object 
 12  TechSupport       7021 non-null   object 
 13  StreamingTV       7021 non-null   object 
 14  StreamingMovies   7021 non-null   object 
 15  Contract          7021 non-null   object 
 16  PaperlessBilling  7021 non-null   object 
 17  

# 📊 EDA

## 1. 🔍 Verificação Colunas

In [9]:
for col in df.columns:
  print(f'{col}: {df[col].unique()}')
  if df[col].nunique() < 50:
    print(df[col].unique())
    print('-' * 50)

Churn: ['No' 'Yes']
['No' 'Yes']
--------------------------------------------------
Gender: ['Female' 'Male']
['Female' 'Male']
--------------------------------------------------
SeniorCitizen: ['No' 'Yes']
['No' 'Yes']
--------------------------------------------------
Partner: ['Yes' 'No']
['Yes' 'No']
--------------------------------------------------
Dependents: ['Yes' 'No']
['Yes' 'No']
--------------------------------------------------
Tenure: [ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 37 49 66 67
 20 43 59 12 27  2 25 29 14 35 64 39 40 11  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 31 36 17 18 51 38 42
  0]
PhoneService: ['Yes' 'No']
['Yes' 'No']
--------------------------------------------------
MultipleLines: ['No' 'Yes']
['No' 'Yes']
--------------------------------------------------
InternetService: ['DSL' 'Fiber optic' 'No']
['DSL' 'Fiber optic' 'No']
--------------------------------------------------
OnlineSecurity: [

## 2. Valores Duplicados

In [10]:
print('Número de Valores Duplicados', df.duplicated().sum())

Número de Valores Duplicados 0


## 3. Valores Nulos

In [11]:
print('Número de Valores Nulos \n', df.isnull().sum())

Número de Valores Nulos 
 Churn               0
Gender              0
SeniorCitizen       0
Partner             0
Dependents          0
Tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
ChargesDaily        0
ChargesMonthly      0
ChargesTotal        0
dtype: int64


## 4. Valores Vazios ou em Branco

In [12]:
df.apply(lambda x: x.astype(str).str.strip() == '').sum()

Churn               0
Gender              0
SeniorCitizen       0
Partner             0
Dependents          0
Tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
ChargesDaily        0
ChargesMonthly      0
ChargesTotal        0
dtype: int64

## 5. 📊 Carga e Verificação

### 5.1. Verificação Variável Alvo

In [13]:
import telecom_x_analytics as tax


fig = tax.plot_distribuicao_target(df)
fig.show()

* **Churn (Geral):** A base apresenta **26,4%** de evasão. Este indicador serve como o *Baseline* (ponto de partida) para compararmos se os subgrupos analizados abaixo estão acima ou abaixo da média da empresa.

### 5.2. Verificação Variáveis Categóricas

#### Análise Perfil Demográfico

In [14]:
import telecom_x_analytics as tax

fig = tax.plot_perfil_demografico(df)
fig.show()

* **Gender (Gênero):** A taxa de Churn é equilibrada entre homens e mulheres, indicando que o gênero não é um fator determinante para a saída.
* **Senior Citizen (Idosos):** Clientes idosos apresentam uma propensão ao Churn significativamente maior que os mais jovens.
* **Partner (Parceiro):** Clientes que não possuem parceiros/cônjuges tendem a cancelar mais o serviço.
* **Dependents (Dependentes):** A ausência de dependentes está fortemente ligada a uma maior taxa de evasão; clientes com família demonstram maior estabilidade.

#### Análise Serviços Infraestrutura

In [15]:
import telecom_x_analytics as tax

fig = tax.plot_servicos_infra(df)
fig.show()

* **Phone Service:** A presença ou ausência de serviço telefônico isolado não altera drasticamente o Churn.
* **Multiple Lines:** Ter múltiplas linhas apresenta uma taxa de evasão ligeiramente maior que linha única.
* **Internet Service:** O ponto crítico; clientes com **Fibra Óptica** têm as maiores taxas de Churn, enquanto clientes com DSL são mais leais.

#### Análise Serviços Internet (Segurança)

In [16]:
import telecom_x_analytics as tax

fig = tax.plot_servicos_seg(df)
fig.show()

* **Online Security:** Clientes sem segurança online saem muito mais. O serviço funciona como um forte retentor.
* **Online Backup:** A ausência de backup está correlacionada a uma maior rotatividade.
* **Device Protection:** Clientes sem proteção de dispositivo apresentam maior índice de cancelamento.
* **Tech Support:** A falta de suporte técnico é um dos maiores preditores de Churn nesta categoria.

#### Análise Serviços Internet (Entretenimento)

In [17]:
import telecom_x_analytics as tax

fig = tax.plot_servicos_entret(df)
fig.show()

* **Streaming TV & Streaming Movies:** Ambos apresentam comportamentos similares; clientes que utilizam esses serviços possuem um Churn moderado, mas acima de quem não possui internet, devido ao valor agregado da fatura.

#### Aspectos Contratuais e Financeiros

In [18]:
import telecom_x_analytics as tax

fig = tax.plot_perfil_contratual_financeiro(df)
fig.show()

* **Contract:** O contrato **Month-to-month** (mensal) detém a esmagadora maioria dos cancelamentos. Contratos de 1 ou 2 anos retêm o cliente com sucesso.
* **Paperless Billing:** Clientes que optam pela fatura digital (sem papel) curiosamente apresentam maior Churn.
* **Payment Method:** O **Electronic Check** (Boleto Eletrônico) é o método com maior índice de evasão. Métodos automáticos (Cartão/Débito) são os mais estáveis.

### 5.3. Verificação Variáveis Numéricas

In [19]:
import telecom_x_analytics as tax

fig = tax.plot_boxplots_numericos(df)
fig.show()

* **Tenure:** A mediana de tempo de casa de quem sai é drasticamente inferior (clientes novos).
* **Charges Monthly:** O Churn está concentrado em valores mensais mais altos (mediana superior).
* **Charges Daily:** Segue o padrão da mensalidade, reforçando o impacto do custo diário na decisão de saída.
* **Charges Total :** Mostra que o Churn ocorre em faturamentos acumulados baixos, antes do cliente se tornar lucrativo.

#### Verificaçao Outliers

In [20]:
import telecom_x_analytics as tax

# Para ver APENAS QUEM DEU CHURN:
tax.exibir_analise_outliers(df, apenas_churn=True)

Variável,Qtd de Outliers,Percentual (%),Limite Inf,Limite Sup
Tenure,23,1.24%,-38.50,69.50
ChargesDaily,0,0.00%,-0.01,5.03
ChargesMonthly,0,0.00%,-0.12,150.88
ChargesTotal,106,5.71%,-3166.05,5651.95


In [21]:
import telecom_x_analytics as tax

# Para ver o dataset COMPLETO:
tax.exibir_analise_outliers(df)

Variável,Qtd de Outliers,Percentual (%),Limite Inf,Limite Sup
Tenure,0,0.00%,-60.00,124.00
ChargesDaily,0,0.00%,-1.52,5.72
ChargesMonthly,0,0.00%,-45.48,171.12
ChargesTotal,0,0.00%,-4694.17,8899.22


#### Análise Comparativa Tendência Numéricas

In [22]:
import telecom_x_analytics as tax

cols_num = ['Tenure', 'ChargesMonthly', 'ChargesTotal']
fig = tax.plot_num_eda_kde(df, cols_num, "EDA: Análise Comparativa de Tendências Numéricas")

fig.show()

* **Tenure:** Pico de probabilidade de Churn nos meses 0 a 6.
* **Charges Monthly:** Aumento da densidade de cancelamento conforme o valor da conta ultrapassa a faixa dos $70.
* **Charges Total:** A curva de Churn decai rapidamente conforme o valor total acumulado aumenta.

#### Matriz de Correlação (Variáveis Numéricas)

In [23]:
import telecom_x_analytics as tax

# Gera e exibe o Heatmap
fig = tax.plot_heatmap_correlacao(df)
fig.show()

* **Insight:** Forte correlação positiva entre **`ChargesMonthly`** e **`ChargesDaily`** (redundância). Observamos também que o **`Tenure`** tem correlação negativa com o Churn (quanto mais tempo de casa, menor a chance de sair).
* **Análise Técnica:** A matriz revela uma colinearidade crítica entre **`ChargesMonthly`**, **`ChargesDaily`** e **`ChargesTotal`**. 
* **Ação de Pré-processamento:** As variáveis de custo diário e faturamento total foram marcadas para exclusão antes do *split* da base.
* **Justificativa:** Manter variáveis numericamente redundantes infla artificialmente o peso desses atributos no modelo e prejudica algoritmos que dependem da independência entre as variáveis. A remoção garante um modelo mais leve e estatisticamente confiável.


#### Teste Qui-Quadrado ($\chi^2$)

In [24]:
import telecom_x_analytics as tax

tabela_decisao = tax.rank_categoricas_completo(df)
display(tabela_decisao)

,Variável,Chi-Score,P-Value,Importância,Manter na Pipeline?
13,Contract,1179.80,6.442285e-257,Alta,Sim
6,InternetService,732.06,1.085779e-159,Alta,Sim
15,PaymentMethod,643.92,3.026775e-139,Alta,Sim
14,PaperlessBilling,254.96,2.151898e-57,Alta,Sim
7,OnlineSecurity,203.30,3.984229e-46,Alta,Sim
10,TechSupport,187.87,9.254887e-43,Alta,Sim
3,Dependents,186.79,1.599568e-42,Alta,Sim
1,SeniorCitizen,160.47,8.934159e-37,Alta,Sim
2,Partner,155.48,1.099962e-35,Alta,Sim
8,OnlineBackup,45.78,1.320621e-11,Alta,Sim


* **Observação:** Esta tabela consolida o cruzamento dos dados antes da fase de treinamento. Note que, além da exclusão das numéricas citadas anteriormente, houve a retirada estratégica de variáveis categóricas redundantes.
* **Por que retirar categóricas específicas?** 1. **Evitar a Armadilha da Variável Dummy (Dummy Variable Trap):** Para variáveis binárias (como Gênero ou Sim/Não), uma coluna é capaz de representar as duas informações. Manter ambas causa redundância perfeita.
    2. **Redução de Dimensionalidade:** Ao remover categorias que não trazem informação nova ou que são constantes, reduzimos o ruído para os modelos de árvore (como XGBoost e Random Forest), permitindo que eles foquem apenas nos atributos que realmente explicam a variação do Churn.
* **Resultado:** O modelo recebe uma base de dados otimizada, o que acelera o treinamento e melhora a interpretabilidade dos resultados.

### Remoção das Colunas para Split

In [25]:
colunas_remove = ['Gender', 'PhoneService', 'ChargesDaily', 'ChargesTotal']
df_clean = df.drop(columns=colunas_remove, errors='ignore')

# 2. Verificar novos duplicados (A sua teoria na prática!)
# Como removemos 'gender' e 'PhoneService', algumas linhas podem ter ficado idênticas
print(f"Duplicados encontrados após o drop: {df_clean.duplicated().sum()}")

# 3. Limpar os duplicados
df_clean = df_clean.drop_duplicates()

# 4. Checagem final do shape
print(f"Checagem de valores duplicados: {df_clean.duplicated().sum()}")
print(f"Shape final do dataset: {df_clean.shape}")

Duplicados encontrados após o drop: 36
Checagem de valores duplicados: 0
Shape final do dataset: (6985, 17)


In [26]:
df_clean.head()

,Churn,SeniorCitizen,Partner,Dependents,Tenure,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,ChargesMonthly
0,No,No,Yes,Yes,9,No,DSL,No,Yes,No,Yes,Yes,No,One year,Yes,Mailed check,65.6
1,No,No,No,No,9,Yes,DSL,No,No,No,No,No,Yes,Month-to-month,No,Mailed check,59.9
2,Yes,No,No,No,4,No,Fiber optic,No,No,Yes,No,No,No,Month-to-month,Yes,Electronic check,73.9
3,Yes,Yes,Yes,No,13,No,Fiber optic,No,Yes,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,98.0
4,Yes,Yes,Yes,No,3,No,Fiber optic,No,No,No,Yes,Yes,No,Month-to-month,Yes,Mailed check,83.9


In [27]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6985 entries, 0 to 7042
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Churn             6985 non-null   object 
 1   SeniorCitizen     6985 non-null   object 
 2   Partner           6985 non-null   object 
 3   Dependents        6985 non-null   object 
 4   Tenure            6985 non-null   int64  
 5   MultipleLines     6985 non-null   object 
 6   InternetService   6985 non-null   object 
 7   OnlineSecurity    6985 non-null   object 
 8   OnlineBackup      6985 non-null   object 
 9   DeviceProtection  6985 non-null   object 
 10  TechSupport       6985 non-null   object 
 11  StreamingTV       6985 non-null   object 
 12  StreamingMovies   6985 non-null   object 
 13  Contract          6985 non-null   object 
 14  PaperlessBilling  6985 non-null   object 
 15  PaymentMethod     6985 non-null   object 
 16  ChargesMonthly    6985 non-null   float64
dtype

# ➗ Split (Divisão de Treino e Teste)

### 1. Separaçao de X e y

In [28]:
# 1. Primeiro: Separa o que é pergunta (X) do que é resposta (y)
X = df_clean.drop(columns=['Churn'])
y = df_clean['Churn']

In [29]:
X.head()

,SeniorCitizen,Partner,Dependents,Tenure,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,ChargesMonthly
0,No,Yes,Yes,9,No,DSL,No,Yes,No,Yes,Yes,No,One year,Yes,Mailed check,65.6
1,No,No,No,9,Yes,DSL,No,No,No,No,No,Yes,Month-to-month,No,Mailed check,59.9
2,No,No,No,4,No,Fiber optic,No,No,Yes,No,No,No,Month-to-month,Yes,Electronic check,73.9
3,Yes,Yes,No,13,No,Fiber optic,No,Yes,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,98.0
4,Yes,Yes,No,3,No,Fiber optic,No,No,No,Yes,Yes,No,Month-to-month,Yes,Mailed check,83.9


In [30]:
y.head()

0     No
1     No
2    Yes
3    Yes
4    Yes
Name: Churn, dtype: object

In [31]:
y.value_counts(normalize=True) * 100

Churn
No     73.586256
Yes    26.413744
Name: proportion, dtype: float64

### 2. Split

In [32]:
# 2. Segundo: Divide em Treino e Teste (O Split propriamente dito)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, 
    stratify=y,
    test_size=0.2, 
    random_state=42)

print("Separação e Split concluídos com sucesso!")

Separação e Split concluídos com sucesso!


# 🛠️ Pré Processamento

In [33]:
from sklearn.compose import ColumnTransformer, make_column_selector
from sklearn.preprocessing import StandardScaler, OneHotEncoder

# Criando o pré-processador com FILTROS AUTOMÁTICOS
pre_processamento = ColumnTransformer(
    transformers=[
        # Filtra automaticamente colunas numéricas (float64, int64)
        ('num', StandardScaler(), make_column_selector(dtype_include=['number'])),
        
        # Filtra automaticamente colunas categóricas (object, category)
        ('cat', OneHotEncoder(drop='first', handle_unknown='ignore'), make_column_selector(dtype_exclude=['number']))
    ],
    verbose_feature_names_out=False
)

# Agora é só aplicar nos dados que vieram do Split!
X_train_transformed = pre_processamento.fit_transform(X_train)
X_test_transformed = pre_processamento.transform(X_test)

### 🔢 Avaliação Pré Processamento

In [34]:
import numpy as np


print(f"Colunas antes: {X_train.shape[1]}")
print(f"Colunas depois: {X_train_transformed.shape[1]}")


print(f"Média das colunas: {np.mean(X_train_transformed):.2f}") 
# Esperado: 0.00 (ou algo como 1.23e-16, que é quase zero)

print(f"Desvio padrão médio: {np.std(X_train_transformed):.2f}")
# Esperado: Próximo de 1.00b


print(f"Valor mínimo: {X_train_transformed.min():.2f}")
print(f"Valor máximo: {X_train_transformed.max():.2f}")

Colunas antes: 16
Colunas depois: 20
Média das colunas: 0.30
Desvio padrão médio: 0.56
Valor mínimo: -1.56
Valor máximo: 1.79


### 🔍 Dataset obtido para Treinamento

In [35]:
# 1. Pegar os nomes das colunas que o pré-processador gerou
# (O método get_feature_names_out é a chave mágica aqui)
nomes_colunas = pre_processamento.get_feature_names_out()

# 2. Criar o DataFrame usando esses nomes
df_treinado_visual = pd.DataFrame(X_train_transformed, columns=nomes_colunas)

# 3. Ver o resultado
df_treinado_visual.head()

,Tenure,ChargesMonthly,SeniorCitizen_Yes,Partner_Yes,Dependents_Yes,MultipleLines_Yes,InternetService_Fiber optic,InternetService_No,OnlineSecurity_Yes,OnlineBackup_Yes,DeviceProtection_Yes,TechSupport_Yes,StreamingTV_Yes,StreamingMovies_Yes,Contract_One year,Contract_Two year,PaperlessBilling_Yes,PaymentMethod_Credit card (automatic),PaymentMethod_Electronic check,PaymentMethod_Mailed check
0,-0.842266,-1.468788,0.0,1.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0
1,0.337940,1.643551,0.0,0.0,0.0,1.0,1.0,0.0,1.0,1.0,1.0,1.0,1.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0
2,-1.289930,0.306361,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0
3,0.826301,1.170622,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,1.0,1.0,0.0,1.0,0.0,0.0,1.0
4,-0.516691,0.283047,1.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0


# 🎯 Treinamento 

### Modelo ⬆️ - Regressão Logística

In [36]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

# 1. Treinamento dos Modelos
# Sem ajuste
log_simples = LogisticRegression(max_iter=1000, random_state=42)
log_simples.fit(X_train_transformed, y_train)

# Com ajuste de balanceamento
log_bal = LogisticRegression(max_iter=1000, class_weight='balanced', random_state=42)
log_bal.fit(X_train_transformed, y_train)

# 2. Gerar e Armazenar as Previsões (A Dica: salvar para reutilizar)
y_pred_log_simples = log_simples.predict(X_test_transformed)
y_pred_log_bal = log_bal.predict(X_test_transformed)

### 🔢 Avaliação Performance - Regressão Logística

In [37]:
import telecom_x_analytics as tax


tax.display_side_by_side(y_test, y_pred_log_simples, y_pred_log_bal, "Regressão Logística")

,precision,recall,f1-score,support
No,0.85,0.90,0.87,1028.00
Yes,0.67,0.54,0.60,369.00
ACURÁCIA GERAL,0.81,0.81,0.81,0.81
,precision,recall,f1-score,support
No,0.90,0.74,0.81,1028.00
Yes,0.52,0.78,0.62,369.00
ACURÁCIA GERAL,0.75,0.75,0.75,0.75


* **Observação:** Avaliamos as métricas de Precision, Recall e F1-Score para o modelo base. Esta tabela serve como referência inicial para entender o potencial de previsibilidade linear dos dados da TelecomX.

### 📉 Matriz Confusão - Regressão Logística

In [38]:
import telecom_x_analytics as tax

fig = tax.plot_comparacao_matrizes(y_test, y_pred_log_simples, y_pred_log_bal, "Regressão Logística")

fig.show()

* **Observação:** O modelo Balanceado é preferível para a TelecomX porque, embora gere mais 'alarmes falsos' (268 vs 100), ele reduz drasticamente o número de clientes que sairiam sem que percebêssemos (cai de 168 para apenas 81 falhas).

### 📉 Gráfico Coeficientes - Regressão Logística

In [39]:
import telecom_x_analytics as tax

# Para o modelo balanceado
fig_coef = tax.plot_coeficientes_regressao(log_bal, pre_processamento)
fig_coef.show()

* **Observação:** Este gráfico detalha o peso e a direção (positiva ou negativa) que cada variável exerce na probabilidade de Churn. É fundamental para validar estatisticamente os insights da análise exploratória.

### 📌 Validação Cruzada - Regressão Logística

In [40]:
from sklearn.model_selection import cross_val_score

# Testando a acurácia em 5 'dobras' diferentes dos dados
scores = cross_val_score(log_bal, X_train_transformed, y_train, cv=5, scoring='accuracy')

print(f"--- VALIDAÇÃO CRUZADA (5-FOLD) ---")
print(f"Acurácia Média: {scores.mean():.2%}")
print(f"Desvio Padrão: {scores.std():.2%}")

if scores.std() < 0.05:
    print("✅ Modelo Robusto: A variação entre os testes é baixa.")
else:
    print("⚠️ Modelo Instável: A performance muda muito dependendo dos dados.")

--- VALIDAÇÃO CRUZADA (5-FOLD) ---
Acurácia Média: 74.75%
Desvio Padrão: 0.93%
✅ Modelo Robusto: A variação entre os testes é baixa.


### Modelo 2️⃣ - Random Forest

In [41]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

# 1. Treinamento dos Modelos de Floresta Aleatória
# Sem ajuste
rf_simples = RandomForestClassifier(n_estimators=100, random_state=42)
rf_simples.fit(X_train_transformed, y_train)

# Com ajuste de balanceamento
rf_bal = RandomForestClassifier(n_estimators=100, class_weight='balanced', random_state=42)
rf_bal.fit(X_train_transformed, y_train)

# 2. Gerar e Armazenar as Previsões
y_pred_rf_simples = rf_simples.predict(X_test_transformed)
y_pred_rf_bal = rf_bal.predict(X_test_transformed)

### 🔢 Avaliação Performance - Random Forest

In [42]:
import telecom_x_analytics as tax


tax.display_side_by_side(y_test, y_pred_rf_simples, y_pred_rf_bal, "Random Forest")

,precision,recall,f1-score,support
No,0.82,0.90,0.85,1028.00
Yes,0.60,0.44,0.51,369.00
ACURÁCIA GERAL,0.78,0.78,0.78,0.78
,precision,recall,f1-score,support
No,0.81,0.90,0.85,1028.00
Yes,0.60,0.43,0.50,369.00
ACURÁCIA GERAL,0.77,0.77,0.77,0.77


* **Observação:** Verificamos a performance do modelo de florestas aleatórias. Esperamos uma evolução nas métricas em relação ao modelo linear ao capturar relações mais complexas entre os serviços.

### 📉 Matriz Confusão - Random Forest

In [43]:
import telecom_x_analytics as tax

fig = tax.plot_comparacao_matrizes(y_test, y_pred_rf_simples, y_pred_rf_bal, "Random Forest")
fig.show()

* **Observação:** Diferente da Regressão Logística, a Random Forest mostrou-se insensível ao balanceamento de classes, mantendo um alto número de Falsos Negativos (210). Para o objetivo de salvar o máximo de clientes, este modelo é inferior à Regressão Logística Balanceada (que falhou em apenas 81 casos).

### 📉 Gráfico Feature Importance - Random Forest

In [44]:
import telecom_x_analytics as tax

fig = tax.plot_feat_importance_rf(rf_simples, nomes_colunas)
fig.show()

* **Observação:** O modelo destaca quais características foram mais decisivas para as "árvores de decisão". Variáveis como Tenure e Mensalidade costumam ganhar destaque nesta fase.

### Modelo 3️⃣ - XGBoost

In [45]:
# !pip install xgboost

In [46]:
# Convertendo o alvo para numérico (0 e 1)
y_train_xgb = y_train.map({'No': 0, 'Yes': 1})
y_test_xgb = y_test.map({'No': 0, 'Yes': 1})

In [47]:
from xgboost import XGBClassifier
from sklearn.metrics import classification_report, confusion_matrix

# Criando o modelo
xgb_model = XGBClassifier(
    n_estimators=100,
    max_depth=4,
    learning_rate=0.1,
    scale_pos_weight=3, # Balanceamento das classes
    random_state=42,
    eval_metric='logloss'
)

# Treinando
xgb_model.fit(X_train_transformed, y_train_xgb)

# Previsões
y_pred_xgb = xgb_model.predict(X_test_transformed)

### 🔢 Avaliação Performance - XGBoost

In [48]:
import telecom_x_analytics as tax

tax.display_side_by_side_xgb(y_test_xgb, y_pred_xgb, "XGBoost")

,precision,recall,f1-score,support
No,0.91,0.73,0.81,1028.00
Yes,0.52,0.80,0.63,369.00
ACURÁCIA GERAL,0.75,0.75,0.75,0.75


* **Observação:** Análise detalhada da performance do algoritmo de Gradient Boosting. Este modelo é conhecido por sua alta eficiência em problemas de classificação binária desbalanceados.

### 📈 Matriz Confusão - XGBoost

In [49]:
import telecom_x_analytics as tax

fig = tax.plot_matriz_single_xgboost(y_test_xgb, y_pred_xgb, "XGBoost")
fig.show()

* **Observação:** O modelo XGBoost consolidou-se como a melhor solução para o problema de Churn da TelecomX. Enquanto a Random Forest apresentou dificuldades em lidar com o desbalanceamento das classes, o XGBoost, aliado à técnica de SMOTE (ou balanceamento), conseguiu capturar 296 dos 369 casos reais de Churn no conjunto de teste.

### 📌 Score AUC - XGBoost

In [50]:
from sklearn.metrics import roc_auc_score

# Calculando a probabilidade (necessário para o AUC)
y_proba_xgb = xgb_model.predict_proba(X_test_transformed)[:, 1]
auc_xgb = roc_auc_score(y_test_xgb, y_proba_xgb)

print(f"O AUC do XGBoost é: {auc_xgb:.4f}")

O AUC do XGBoost é: 0.8436


### 📈 Gráfico Feature Importance - XGBoost

In [51]:
import telecom_x_analytics as tax

fig = tax.plot_feat_importance_xg(xgb_model, nomes_colunas)
fig.show()

* **Observação:** Diferente dos outros modelos, o XGBoost 'entendeu' que a fidelização contratual é muito mais forte que o valor da mensalidade para segurar o cliente na base.

### Modelo 4️⃣ - SVM

In [52]:
from sklearn.svm import SVC

# 1. Treinamento dos Modelos
# Sem ajuste
svm_simples = SVC(kernel='rbf', probability=True, random_state=42)
svm_simples.fit(X_train_transformed, y_train)


# Com ajuste de balanceamento
svm_bal = SVC(kernel='rbf', class_weight='balanced', probability=True, random_state=42)
svm_bal.fit(X_train_transformed, y_train)

# 2. Gerar e Armazenar as Previsões (A Dica: salvar para reutilizar)
y_pred_svm_simples = svm_simples.predict(X_test_transformed)
y_pred_svm_bal = svm_bal.predict(X_test_transformed)

### 🔢 Avaliação Performance - SVM

In [53]:
import telecom_x_analytics as tax


tax.display_side_by_side(y_test, y_pred_svm_simples, y_pred_svm_bal, "Support Vector Machine (SVM)")

,precision,recall,f1-score,support
No,0.84,0.91,0.87,1028.0
Yes,0.66,0.50,0.57,369.0
ACURÁCIA GERAL,0.80,0.80,0.80,0.8
,precision,recall,f1-score,support
No,0.89,0.74,0.81,1028.00
Yes,0.50,0.74,0.60,369.00
ACURÁCIA GERAL,0.74,0.74,0.74,0.74


* **Observação:** Avaliamos a eficácia da separação por hiperplanos. Verificamos se este modelo consegue superar os algoritmos baseados em árvores em termos de precisão e generalização.

### 📉 Matriz Confusão - SVM

In [54]:
import telecom_x_analytics as tax

fig = tax.plot_comparacao_matrizes(y_test, y_pred_svm_simples, y_pred_svm_bal, "SVM")
fig.show()

* **Observação:** O modelo SVM (Support Vector Machine) apresentou um comportamento muito similar à Regressão Logística quando balanceado. Ele conseguiu reduzir significativamente as falhas de detecção (Falsos Negativos), caindo de 161 para 95 casos.

### 📝 Comparação dos 4 Modelos

In [55]:
import telecom_x_analytics as tax

# Criamos a lista de tuplas (predição, "nome")
modelos_para_comparar = [
    (y_pred_log_bal, "Regressão Logística"),
    (y_pred_rf_bal, "Random Forest"),
    (y_pred_xgb, "XGBoost"),
    (y_pred_svm_bal, "SVM")
]

# Uma única linha gera o ranking completo e estilizado
ranking = tax.exibir_ranking_modelos(modelos_para_comparar, y_test)

,Acurácia,Recall (Churn),Precisão (Churn),F1-Score
Modelo,,,,
XGBoost,75.16%,80.22%,51.93%,63.05%
Regressão Logística,75.02%,78.05%,51.80%,62.27%
SVM,73.87%,74.25%,50.37%,60.02%
Random Forest,77.45%,43.09%,60.23%,50.24%


* **Observação:** O **XGBoost** venceu o ranking oficial da TelecomX, apresentando a melhor métrica de AUC e estabilidade em dados não vistos.

### 📈 Curva ROC

In [56]:
import telecom_x_analytics as tax

# 1. Lista de tuplas (Nome, Objeto_do_Modelo)
modelos_comp = [
    ("Regressão Logística", log_bal),
    ("Random Forest", rf_bal),
    ("XGBoost", xgb_model),
    ("SVM", svm_bal)
]

# 2. Chamada única e limpa
fig = tax.plot_roc_comparativa(modelos_comp, X_test_transformed, y_test)
fig.show()

* **Observação:** O modelo atingiu um **AUC de 0.844**. Isso significa que em 84% das vezes, o modelo classificará corretamente um cliente de alto risco acima de um cliente estável.

# ⚙️ Artefatos do Modelo

## ➜] Exportação do Modelo Preditivo Final (XGBoost)

In [57]:
import pickle

In [58]:
try:
    with open('modelo_churn_xgb.pkl', 'wb') as file:
        pickle.dump(xgb_model, file)
    print('Modelo salvo com sucesso!!!')
except Exception as e:
    print('Ocorreu um erro ao salvar o modelo: ', str(e))

Modelo salvo com sucesso!!!


In [59]:
# Teste rápido de carregamento
try:
    with open('modelo_churn_xgb.pkl', 'rb') as file:
        modelo_teste = pickle.load(file)
    print('Modelo carregado e validado com sucesso!')
except Exception as e:
    print('Erro ao validar o arquivo gerado: ', str(e))

Modelo carregado e validado com sucesso!


## ➜] Exportação do Objeto de Pré-processamento

In [60]:
try:
    with open('pre_processador.pkl', 'wb') as file:
        pickle.dump(pre_processamento, file)
    print('Modelo salvo com sucesso!!!')
except Exception as e:
    print('Ocorreu um erro ao salvar o modelo: ', str(e))

Modelo salvo com sucesso!!!


In [61]:
# Teste rápido de carregamento
try:
    with open('pre_processador.pkl', 'rb') as file:
        modelo_teste = pickle.load(file)
    print('Modelo carregado e validado com sucesso!')
except Exception as e:
    print('Erro ao validar o arquivo gerado: ', str(e))

Modelo carregado e validado com sucesso!


# >

<style>
    /* Cores Modo Claro */
    :root {
        --cor-texto: #333333;
        --cor-titulo: #3498db;
        --cor-bg-box: #f1f7fc;
    }

    /* Cores Modo Escuro */
    @media (prefers-color-scheme: dark) {
        :root {
            --cor-texto: #e0e0e0;
            --cor-titulo: #5dade2;
            --cor-bg-box: #1c2a35;
        }
    }

    p, li { font-size: 18px !important; color: var(--cor-texto) !important; line-height: 1.6; }
    h1 { font-size: 34px !important; color: var(--cor-titulo) !important; border-bottom: 2px solid var(--cor-titulo); padding-bottom: 10px; }
    h2 { font-size: 28px !important; color: var(--cor-titulo) !important; }

    /* Caixa de Análise */
    .analise-box {
        font-style: italic;
        border-left: 6px solid var(--cor-titulo) !important;
        background-color: var(--cor-bg-box) !important;
        color: var(--cor-texto) !important;
        padding: 15px;
        margin: 20px 0;
        display: block;
    }
    .analise-box strong { color: var(--cor-titulo); }

    /* Efeito de Zoom Direcional */
    .zoom-right, .zoom-left { transition: all 0.4s ease; cursor: zoom-in; position: relative; z-index: 1; }
    .zoom-right:hover { transform: scale(1.8); transform-origin: left center; z-index: 9999; box-shadow: 0 10px 30px rgba(0,0,0,0.5); }
    .zoom-left:hover { transform: scale(1.8); transform-origin: right center; z-index: 9999; box-shadow: 0 10px 30px rgba(0,0,0,0.5); }
    
    table, td, th { overflow: visible !important; border: none !important; }
    table { width: 100%; border-collapse: separate !important; border-spacing: 15px !important; }
</style>

# 📑 Relatório de Análise Preditiva: Estratégias de Churn TelecomX

## 1. Sumário Executivo
Este documento apresenta os resultados da pipeline de dados desenvolvida para a **TelecomX**, focada na antecipação do cancelamento de clientes `Churn`. O objetivo é transformar dados históricos em ações preventivas para garantir a sustentabilidade da receita.

### Panorama da Base de Clientes
<br/>
<img src="https://raw.githubusercontent.com/WBruni/analise-churn-telecom-x_parte_2/main/images/EDA_Distribuicao_Churn.png" width="60%">

* **Diagnóstico:** Identificamos que **26,4%** da nossa base de clientes cancelou o serviço recentemente. Este volume representa uma perda significativa de receita recorrente e justifica a implementação de uma estratégia de retenção baseada em dados.

---

## 2. Diagnóstico do Comportamento de Saída (Insights de Negócio)
A análise detalhada revelou que o Churn na TelecomX segue padrões comportamentais e financeiros claros.

### Perfil do Cliente em Risco
<br/>
<img src="https://raw.githubusercontent.com/WBruni/analise-churn-telecom-x_parte_2/main/images/EDA-Perfil_Demografico.png" width="80%">

* **Fatores Demográficos:** Observamos uma maior vulnerabilidade em clientes **seniores** e naqueles que **não possuem dependentes**. Estes perfis demandam uma comunicação mais personalizada e ofertas que tragam segurança e praticidade.

### Análise de Atrito Contratual e Financeiro
<br/>
<img src="https://raw.githubusercontent.com/WBruni/analise-churn-telecom-x_parte_2/main/images/EDA-Contrato_Faturamento.png" width="80%">

* **Vulnerabilidade Contratual:** O modelo de **Contrato Mensal** é o maior facilitador da saída. Além disso, o **Boleto Eletrônico** apresenta índices de Churn drasticamente superiores aos métodos automáticos (Cartão e Débito).

---

## 3. Metodologia e Inteligência Preditiva
Para a construção da solução, a pipeline passou por um rigoroso processo de seleção de atributos, focado em eliminar o "ruído" estatístico e garantir a estabilidade das previsões.

### Refinação e Estabilidade Numérica (Matriz de Correlação)
Abaixo, a matriz ilustra o tratamento de redundâncias. Variáveis com altíssima colinearidade foram tratadas para evitar o sobreajuste (overfitting), garantindo que cada atributo numérico traga uma informação única para o algoritmo.
<br/>
<img src="https://raw.githubusercontent.com/WBruni/analise-churn-telecom-x_parte_2/main/images/EDA_Matriz_Correlacao.png" width="70%">

### Lógica de Seleção Categórica (Tabela de Decisão)
Para as variáveis qualitativas, aplicamos uma **Tabela de Decisão**. Este método permitiu converter o comportamento dos serviços (como Fibra Óptica e Suporte Técnico) em pesos lógicos, facilitando o aprendizado dos modelos e tornando a predição mais assertiva.
<br/>
<img src="https://raw.githubusercontent.com/WBruni/analise-churn-telecom-x_parte_2/main/images/EDA-Tabela_Decisao.png" width="50%">

### Seleção e Performance do Algoritmo
<br/>
<img src="https://raw.githubusercontent.com/WBruni/analise-churn-telecom-x_parte_2/main/images/Ranking_Final_Modelos.png" width="50%">

* **Modelos Avaliados:** Submetemos os dados a quatro algoritmos: **Regressão Logística**, **Random Forest**, **XGBoost** e **SVM**.
* **A Escolha:** O **XGBoost** foi o selecionado devido à sua superioridade em identificar padrões complexos de cancelamento sem gerar um volume excessivo de falsos alertas.

### Confiabilidade do Modelo (Curva ROC)
<br/>
<img src="https://raw.githubusercontent.com/WBruni/analise-churn-telecom-x_parte_2/main/images/Curva_ROC_Modelos.png" width="80%">

* **Poder de Separação:** Com um índice **AUC de 0.844**, o modelo demonstra alta precisão ao distinguir potenciais cancelamentos de clientes satisfeitos.

---

## 4. Direcionamento Estratégico e Plano de Ação

### Importância de Variáveis (XGBoost)
<br/>
<img src="https://raw.githubusercontent.com/WBruni/analise-churn-telecom-x_parte_2/main/images/TR_Importance_XG.png" width="80%">

* **O que define a permanência:** O modelo confirma que o **Tipo de Contrato** e o **Tempo de Casa (Tenure)** são os preditores mais fortes.

### Plano de Ação TelecomX:
1.  **Programa de Estabilização de Contrato:** Migrar clientes "mês a mês" para planos anuais através de benefícios progressivos.
2.  **Otimização Financeira:** Incentivar o uso de métodos de pagamento automático para reduzir o atrito do pagamento manual.
3.  **Auditoria de Experiência em Fibra Óptica:** Investigar a relação entre faturas altas e percepção de qualidade para reduzir o `Churn` de ticket alto.
4.  **Campanha de Valor para Seniores:** Desenvolver pacotes de serviços de suporte e segurança digital focados no público sênior sem dependentes.
5.  **Sistema de Alerta Antecipado (Early Warning):** Integrar o modelo XGBoost ao CRM para sinalizar clientes de "Alto Risco" assim que atingirem o 3º mês de contrato mensal.

---

## 5. Conclusão Geral e Evolução do Projeto
Este relatório consolida a **Parte 2** de um estudo aprofundado sobre a base de clientes da TelecomX. Enquanto a **Parte 1** consistiu em uma análise exploratória (EDA) abrangente para entender o comportamento geral da base, esta **segunda etapa** focou na maturidade preditiva do projeto.

Para alcançar os resultados aqui apresentados, realizamos um refinamento crítico do dataset, eliminando variáveis redundantes ou de baixo impacto identificadas na fase inicial. Esse processo de "limpeza estratégica" permitiu um conjunto de dados mais equilibrado e focado, facilitando o treinamento de modelos mais robustos e precisos. O diagnóstico final é claro: o `Churn` está concentrado em serviços de alto valor e contratos de baixa fidelidade. Com esse plano de ação proposto, a TelecomX está agora equipada não apenas com dados, mas com uma ferramenta de inteligência capaz de proteger sua receita e aumentar a longevidade da relação com seus clientes.